src/preamble.js

```diff
@@ -138,15 +138,17 @@ function ccall(ident, returnType, argTypes, args, opts) {
   }
 #endif
 #if ASYNCIFY && WASM_BACKEND
-  if (typeof Asyncify === 'object' && Asyncify.currData) {
+#if ASSERTIONS
+   // Once the asyncFinalizers are called, asyncFinalizers gets reset to [].
+   // If they are not empty, then another async ccall is in-flight and not finished.
+  assert((!opts || !opts.async) || Asyncify.asyncFinalizers.length === 0, 'Cannot have multiple async ccalls in flight at once');
+#endif
+  if (typeof Asyncify === 'object' && Asyncify.currData && Asyncify.asyncFinalizers.length === 0) {
```

Comment:
Why is `Asyncify.asyncFinalizers.length === 0` added at the end here?

Rereading the code around it, I think this is checking if we are doing an async operation. Could we maybe just check `if (opts && opts.async) {`?

Reply:
The if statement does check if something is doing an async operation:

1. The reason why we can't just use `if (opts && opts.async) {` is because we not only need to see if there is something running with `opts.async` but if it's *truly* asynchronous (i.e. a sleep happens during the execution). Because inside the if-statement we have code for preparing a Promise return value. Later on, if `opts.async === true` and it never slept, we return an immediately resolved Promise. This behavior is analogous the Emterpreter's ccall implementation.

2. The reason why `Asyncify.asyncFinalizers.length === 0` is tacked on at the end is it makes the if statement mean "if we are running asynchronously *and* this is the first asynchronous operation (since the finalizers length is 0)". This is because when follow an async operation with a sync one, if we didn't tack on `Asyncify.asyncFinalizers.length === 0`,  we would enter the if-statement and return a Promise since we would detect something is running asynchronously.

Reply:
I kept the same logic, but re-factored a bit to move the logic into 3 variables that helps it read easier:

1. `asyncMode` - whether the user specified opt.async
2. `runningAsync` - if we are **truly running asynchronously** (something has slept and is currently yielded) during this `ccall`
3. `prevRunningAsync` - if we were previously running **truly asynchronously** before `ccall` happened